sdks/go/pkg/beam/core/runtime/exec/sdf.go

```diff
@@ -647,6 +653,46 @@ func (n *ProcessSizedElementsAndRestrictions) Split(f float64) ([]*FullValue, []
 	return p, r, nil
 }
 
+// Checkpoint splits the remaining work in a restriction into residuals to be resumed
+// later by the runner. This is done iff the underlying Splittable DoFn returns a resuming
+// ProcessContinuation. If the split occurs and the primary restriction is marked as done
+// my the RTracker, the Checkpoint fails as this is a potential data-loss case.
+func (n *ProcessSizedElementsAndRestrictions) Checkpoint() ([]*FullValue, error) {
```

Comment:
Could you please add some test cases? It should be fairly straightforward to model them off of the existing Split tests

Reply:
we've got some unit tests now